air/src/air/transition/frame.rs

```diff
@@ -3,80 +3,81 @@
 // This source code is licensed under the MIT license found in the
 // LICENSE file in the root directory of this source tree.
 
-use super::{FieldElement, Vec};
+use super::{super::Air, FieldElement, Vec};
+use utils::iterators::StreamingIterator;
 
-// EVALUATION FRAME
-// ================================================================================================
 /// A set of execution trace rows required for evaluation of transition constraints.
-///
-/// In the current implementation, an evaluation frame always contains two consecutive rows of the
-/// execution trace. It is passed in as one of the parameters into
+/// It is passed in as one of the parameters into
 /// [Air::evaluate_transition()](crate::Air::evaluate_transition) function.
+pub trait EvaluationFrame<E: FieldElement> {
+    /// Creates an empty frame
+    fn new<A: Air<BaseField = E>>(air: &A) -> Self;
+
+    /// Creates an frame instantiated from the provided rows
+    fn from_rows(rows: Vec<Vec<E>>) -> Self;
+
+    /// Fills the frame using the provided column iterator
+    fn read_from<'a, I: StreamingIterator>(&mut self, columns: I, step: usize);
```

Comment:
Tried this in an attempt to solve the lifetime errors from `fn read_from<'a, I: Iterator<Item = &'a [E]>>(&mut self, columns: I, step: usize)`, but stuck on how to handle the associated type from the iterator.

Reply:
One thought is that maybe instead of passing in an iterator (which needs to be mutable), we pass in something else which provides a view into a 2D structure (which doesn't need to be mutable). For example, let's say we have a `Matrix` trait (a different name is probably needed since we already have `Matrix` struct in the prover) which provides the following interface:

```Rust
pub trait Matrix<E: FieldElement> {

    fn num_cols(&self) -> usize;
    fn num_rows(&self) -> usize;

    fn get(&self, col_idx: usize, row_idx: usize) -> E;
}
```

And then `read_from()` can have the following signature:
```Rust
fn read_from<M: Matrix<E>>(&mut self, data: M, step: usize);
```

And then we'd need to implement `Matrix` trait for w/e struct needs to be passed into `read_from()`.

Reply:
This seems like a cleaner approach. I've called it `TableReader` for now, but am open to better names.  I've also added a `blowup` argument to the signature so that the row indices can be computed for the LDE trace.